File: python/692.前k个高频单词.py
#
# @lc app=leetcode.cn id=692 lang=python3
#
# [692] 前K个高频单词
#
# 解法1(T97% S77%): 这题唯一的不同就是如果次数相同要按字典序输出，所以要有两个排序条件，因为sorted只能从大大小排序而且默认是字典序，因此可以把次数添加一个负号 

# @lc code=start
class Solution:
    def topKFrequent(self, words, k: int):
        table = {}
        for word in words:
            if word in table: table[word] += 1
            else: table[word] = 1
        
        li = sorted(table.items(), key=lambda x: (-x[1], x[0]), reverse=False)

        res = []
        for i in range(k):
            res.append(li[i][0])

        return res
# @lc code=end

File: javascript/692.前k个高频单词.js
/*
 * @lc app=leetcode.cn id=692 lang=javascript
 *
 * [692] 前K个高频单词
 * 
 * 多个排序准则，排序两次就好，先按照第二要求排，再按照第一要求排
 */

// @lc code=start
/**
 * @param {string[]} words
 * @param {number} k
 * @return {string[]}
 */
var topKFrequent = function(words, k) {
    var freq = new Map();
    for (var word of words) {
        if (!freq.has(word)) {
            freq.set(word, 1);
        } else {
            freq.set(word, freq.get(word)+1);
        }
    }
    var freqarr = [...freq.entries()].sort();
    freqarr.sort((a, b) => b[1]-a[1])
    return freqarr.slice(0, k).map(x => x[0]);
};
// @lc code=end

